glommio/src/sync/gate.rs

```diff
@@ -83,17 +83,35 @@ impl Gate {
         )
     }
 
-    /// Close the gate, and wait for all spawned tasks to complete
-    pub async fn close(&self) -> Result<(), GlommioError<()>> {
-        self.inner.close().await
+    /// Close the gate, and return a waiter for all spawned tasks to complete. If the gate is currently closing, the
+    /// returned future will wait for it to close before returning a success. This is particularly useful if you might
+    /// have a timeout on the close - the would otherwise be no safe way to retry & wait for remaining tasks to finish.
+    ///
+    /// NOTE: After this function returns, [is_open](Self::is_open) returns false and any subsequent attempts to acquire
+    /// a pass will fail, even if you drop the future. The future will return an error if and only if the gate is
+    /// already fully closed
+    pub fn close(&self) -> impl Future<Output = Result<(), GlommioError<()>>> {
```

Comment:
Should this be `Future<Output = Result>` or `Result<Output = Future>`? The latter is a cleaner API but the approach taken here means the users of the library don't need to make any code changes when upgrading. On the other hand, they may be silently impacted by a slight behavioral change in the mechanics of the close which may or may not be important.

I went with the "no code change required to upgrade" approach since I figure the behavioral change is unlikely to be noticed but let me know which approach is preferrable.

Reply:
I think what you did is better.

I don't think the latter is a cleaner API, anyway, but since you're not pushing for it, what I think is immaterial =)